### PR TITLE
Adding interCTC loss to hybrid models

### PR DESCRIPTION
**Pull request by Kipok**
# What does this PR do ?

Adding interCTC loss to hybrid models. To solve issue that ctc and hybrid models use different names for things that we need (e.g., wer vs ctc_wer), I'm now directly passing all necessary "main" class variables to the interctc_setup method (except for self.cfg). I'm not sure if there is much sense in still keeping interctc as mixin, given that it does not access anything from the main class anymore, but not making any changes for now.

Additionally, there are 2 slight bugs in the original code. First, inter_ctc_loss was incorrectly calculated in the logs (main loss was used, not inter_ctc there). Note that this is purely a logging thing and does not affect training. Second, main loss was not multiplied by the correct coefficient, which does affect training to some extent. So, the calculation should have been something like `loss = final_loss * 0.7 + inter_loss * 0.3`, while it was just `loss = final_loss + inter_loss * 0.3`

**Collection**: ASR

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/models/hybrid_rnnt_ctc_models.py [hunk 1]
```diff
@@ -441,6 +435,17 @@ def training_step(self, batch, batch_nb):
                 self.ctc_wer.reset()
                 tensorboard_logs.update({'training_batch_wer_ctc': ctc_wer})
 
+        # note that we want to apply interctc independent of whether main ctc
```
Comment (VahidooX) [inline nemo/collections/asr/models/hybrid_rnnt_ctc_models.py:438]: How would be the final loss when ctc_loss_weight=0.3 and interctc.loss_weights=[0.1]? Something like this?
`final_loss=0.9(0.7(rnnt_loss)+3.0(ctc_loss_final_layer))+0.1(ctc_loss_middle_layer)`

File: tests/collections/asr/test_asr_interctc_models.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Comment (VahidooX) [inline tests/collections/asr/test_asr_interctc_models.py:1]: Copyright 2022

File: tests/collections/asr/test_asr_interctc_models.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Dict
+
+import pytest
+import pytorch_lightning as pl
+import torch
+from omegaconf import DictConfig, ListConfig
+
+from nemo.collections.asr.metrics.wer import CTCDecoding, CTCDecodingConfig
```
Comment (VahidooX) [inline tests/collections/asr/test_asr_interctc_models.py:21]: Unused imports here.

**Review by VahidooX — CHANGES_REQUESTED** [review]
LGTM! Some minor comments.

File: nemo/collections/asr/parts/mixins/interctc_mixin.py [hunk 4]
```diff
@@ -189,29 +206,33 @@ def add_interctc_losses(
...
+        loss_value *= self._interctc_params['main_loss_weight']
+
         for layer_idx, intermediate_result, loss_weight in zip(
-            self._apply_at_layers, captured_tensors, self._intermediate_loss_weights
+            self._interctc_params['apply_at_layers'],
+            captured_tensors,
+            self._interctc_params['intermediate_loss_weights'],
         ):
-            inter_loss_value = self.loss(
+            inter_loss_value = self._interctc_params['loss'](
```
Comment (titu1994) [inline nemo/collections/asr/parts/mixins/interctc_mixin.py:219]: Why not use self.loss?

File: nemo/collections/asr/parts/mixins/interctc_mixin.py [hunk 5]
```diff
@@ -189,29 +206,33 @@ def add_interctc_losses(
...
                 targets=transcript,
                 target_lengths=transcript_len,
                 input_lengths=intermediate_result[1],
             )
-            metrics[f"{log_prefix}inter_ctc_loss_l{layer_idx}"] = loss_value.detach()
+            metrics[f"{log_prefix}inter_ctc_loss_l{layer_idx}"] = inter_loss_value.detach()
             loss_value += inter_loss_value * loss_weight
             if compute_wer:
-                self._wer.update(
+                self._interctc_params['wer'].update(
```
Comment (titu1994) [inline nemo/collections/asr/parts/mixins/interctc_mixin.py:228]: Why not use self._wer ? You can check and set self._wer or self.wer (or better yet, add alias of self.wer = self._wer as a property for CTC models.

**Review by titu1994 — CHANGES_REQUESTED** [review]

File: nemo/collections/asr/models/hybrid_rnnt_ctc_models.py (hunk 1 above)
Comment (Kipok) [reply nemo/collections/asr/models/hybrid_rnnt_ctc_models.py:438]: yes, that's correct. I will add an explicit formula here

**Review by VahidooX — DISMISSED** [review]
LGTM! 
Just one not used import of CTCDecoding in test_asr_interctc_models.py.

File: nemo/collections/asr/models/ctc_models.py [hunk 6]
```diff
@@ -104,7 +104,7 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         self.setup_optimization_flags()
 
         # setting up interCTC loss (from InterCTCMixin)
-        self.setup_interctc()
+        self.setup_interctc('decoder', 'loss', '_wer')
```
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:107]: Can you use keyword args here ? Kinda hard to tell what is input to this function

File: nemo/collections/asr/models/hybrid_rnnt_ctc_models.py [hunk 7]
```diff
@@ -88,6 +88,9 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         # setting the RNNT decoder as the default one
         self.use_rnnt_decoder = True
 
+        # setting up interCTC loss (from InterCTCMixin)
+        self.setup_interctc('ctc_decoder', 'ctc_loss', 'ctc_wer')
```
Comment (titu1994) [inline nemo/collections/asr/models/hybrid_rnnt_ctc_models.py:92]: Same as above

**Review by titu1994 — APPROVED** [review]
Looks much better, thanks ! Minor change then let's merge

**Review by titu1994 — APPROVED** [review]
Thanks !
